### ISSUE TITLE
[v2 QUESTION]: How to apply optimized hyperparameters in training?

### ISSUE THREAD

**Issue by llzhu** [chemprop/chemprop#1003]
I am re-writing some applications using chemprop 1.7 and switch to chemprop 2.4. But most predicted results from chemprop 2.4 are not as good as 1.7.

I am thinking to improve the results by optimize the hyperparameters. For a regression data set for bioavailability, I got the following optimal hyperparameters following the steps in hpopting.ipynb:

{'depth': 2, 
 'ffn_hidden_dim': 2200,
 'ffn_num_layers': 2,
 'message_hidden_dim': 400}

[I am a little surprised that it is the same results as in the example provided in hpopting.ipynb; these are 2 totally different data sets with totally different molecules in size and types. Any comments on that?]

I have the following questions:
1. What are the default values for these h-parameters?
2. How to apply these optimized hyperparameters in the training using the steps provided in the example training.ipynb?

Thanks. -Leiming

**Comment by KnathanM** [comment]
> What are the default values for these h-parameters?
In the [message passing](https://chemprop.readthedocs.io/en/main/autoapi/chemprop/nn/message_passing/base/index.html) and [ffn](https://chemprop.readthedocs.io/en/main/autoapi/chemprop/nn/ffn/index.html) classes, you can see the defaults
depth = 3, message_hidden_dim (d_h) = 300, ffn_num_layers = 1, ffn_hidden_dim = 300.
These defaults are a bit on the low side, so usually the best hyperparameters give a larger model. 

> How to apply these optimized hyperparameters in the training using the steps provided in the example training.ipynb?
You can pass the optimized hyperparameters to the model components when creating them. Details of this are documented under "Model Modules" [here](https://chemprop.readthedocs.io/en/main/tutorial/python/index.html). In this case:
```
mp = nn.BondMessagePassing(depth = 2, d_h = 400)
ffn = nn.RegressionFFN(n_layers = 2, hidden_dim=2200)
```
> I am a little surprised that it is the same results as in the example provided in hpopting.ipynb; these are 2 totally different data sets with totally different molecules in size and types. Any comments on that?

I assume that you switched out the data for your data under "Make data points, splits, and datasets. I don't have much of a comment about the hyperparameters being the same except that those values seem reasonable. Maybe try running hpopting again and see if you get different values? The models should be randomly initialized so running it again may give slightly different results. 

> But most predicted results from chemprop 2.4 are not as good as 1.7.
Depending on whether you data is open, we would love to see the details of your data and/or the results you got with 1.7 and 2.4. We have done some benchmarking that we will publish soon where we found that both versions did about equally good. Sometimes v2 would be a little better than v1 and sometimes a little worse. You can also try turning off batch normalization (`MPNN(batch_norm = False)`) which we found helped in some cases.

You can also convert v1 models to v2 models using `convert_v1_to_v2.ipynb` under examples. It may be a bit outdated so please open an issue if you run into any problems with that.

**Comment by llzhu** [comment]
Hi Nathan:
      Thanks for the explanation and suggestions. When using these h-parameters,  
mp = nn.BondMessagePassing(depth = 2, d_h = 400)
ffn = nn.RegressionFFN(n_layers = 2, hidden_dim=2200)

I got the following error message:
mat1 and mat2 shapes cannot be multiplied (64x400 and 300x2200) 
at this line:
trainer.fit(mpnn, train_loader, val_loader)

If I change d_h to 300, it works. It seems that if I use d_h = 400 in mp, I must also change another h-parameter in ffn to 400. It is not hidden_dim. Please advise me.

Thanks. -Leiming

**Comment by KnathanM** [comment]
Sorry I forgot that you also need to change the input dimension of the FFN to match the message passing hidden dimension.
`nn.RegressionFFN(n_layers = 2, hidden_dim=2200, input_dim =400)
